### DISCUSSION TITLE
Dynamic blocked field visibiilty

### DISCUSSION THREAD

**Discussion by knotenpunkt** [graphql-java/graphql-java#2189] [Q&A] [answered]
hey,

how can i dynamically block the field visibility of fields? - dont using more than one graphQL instance.
So that the **introspection** shows if i am logged in as admin me all fields and if i am logged in as a normal user only shows a part of it.

Im not sure if it is possible? But it would be nice.

If it is not possible, do you think it is a good idea, to expose all fields? if i needed as admin or a other role, the introspection view of the fields which should be visible to normal users? What is there actually the best practice?

Maybe you can also combine/integrate the block-field-logic into the directive-logic, so that i can use it directly in my Directvie-Implemenation. Its high cohesive, because mostly i want to restrict not only by the field visibility but also by a directive which then throws a exception if the user tries to access the certain field. (Not only returning null, because thats a semantic-error^^, and a bad design)

Edit: Changed instrumenation through introspection, dont know why i had this big typo-mistake^^

**Comment by bbakerman** [comment, marked as answer]
You have two choices

We have an older `graphql.schema.visibility.GraphqlFieldVisibility` that can dynamically make fields disappear.  They dissapear from introspection and validation but on the same schema instance

if you have a class of users (say admins and non admins) as a dimension then you can use `graphql.schema.SchemaTransformer` once to dynamically create schemas for certain audiences.  You then swap in a schema to use per audience group. Your could run your visitor code based on directives say.

We recommend the multiple schemas per audience class - its more performant since there is no runtime code running to check if a field should be shown or not.  

In fact if we had our time again we would likely have NOT invented `graphql.schema.visibility.GraphqlFieldVisibility` - not that is super slow - but at scale building "targetted schemas" once is more efficient than make decisions on every request

**Reply by knotenpunkt** [reply]
hey,

thank you for your answer, its very helpful.
You are right making deciscions on every request isnt a good idea.

So one re-question:
What you exactly mean with multiple schemas per audience class?
Do you mean there different graphql-instances?
(i know that facebook uses only two graphql instances, one for the extern users for us^^ and one for internal-stuff. At that location i read that, there the people discussed about the approach having only one vs having multiple graphQL instances. The problem is, having more than one you cant reuse some internal parser cache and so on.)

So the question do you mean with multiple schemas multiple graphql instances, and if yes, it is such performant if i would use only one, for a big subset which is equal in the different instances?

Or do you mean with multiple schemas something else, which i dont know right now?^^

**Comment by bbakerman** [comment]
> What you exactly mean with multiple schemas per audience class?


So lets assume you have three classes of users, - say full admins,
project admins and users.

You would create a "schema" that is the superset of all the fields that
everyone can see.

Then you would transform this master schema via `SchemaTransformer` and
create a schema instance (and graphql instance wrapping it) per user set.

Then in your code BEFORE you invoke graphql-java code, you decide what set
the user belongs to.

If they are a full admin, you grab the full admin schema+graphql object and
invoke that.  If they are a plain user you grab a different graphql+schema.

This means you have 3 schemas in memory and 3 wrapping graphql objects.

> The problem is, having more than one you cant reuse some internal parser
cache and so on.

This is true but remember if you "hide fields" then the set of possible
fields changes and caching parsing and so on is fraught because you have
different schemas anyway.

**Reply by knotenpunkt** [reply]
ah ok so the point is, having multiple instances^^

> This is true but remember if you "hide fields" then the set of possible
fields changes and caching parsing and so on is fraught because you have
different schemas anyway.

But i see also your point that this is ok.

I would not only build a superset of alle the fields, but a set of really all fields.
Maybe with the help of directives i would mark fields as blocked.

On the wiring time i would add for example the the normal user the BlockingAdminStuff directive into the wiring-process.

And then i would try to remove it with 

```
@Override
    public GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment)
    {
        String directiveRole = (String) environment.getDirective().getArgument("role").getValue();

        GraphQLFieldDefinition field = environment.getElement();
        GraphQLFieldsContainer parentType = environment.getFieldsContainer();

...

  return field.transform(builder ->{
            builder.???
        });

or return null?

or something like this: https://spectrum.chat/graphql-java/general/what-are-the-correct-and-suggested-ways-to-modify-graphql-schema-dynamically-during-runtime~c698bbb5-0909-4a70-b8f4-81072511a0cf?m=MTU0MzI3ODk2NTc0Ng==

//but in that case i need the parent field which i must tranform?
//this spectrum link is very old, maybe there is a newer nicer solution?


```

offtopic: Maybe you can also take a look on that issue i created in the website-repo: https://github.com/graphql-java/graphql-java-page/issues/45 (its about wrapping data-fetchers in directives, not directly related to this topic here)


for the instance i expose admins i would not add theblocking directive on wiring-time^^

i think that could be a good solution, but with the need to remove fields easily in the directive-java-code

**Comment by bbakerman** [comment]
SchemaDirectiveWiring is not what you want.

Have a look at

graphql.schema.transform.FieldVisibilitySchemaTransformation

and also 

 https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/graphql/schema/SchemaTransformerTest.groovy for and example schema transformation

for examples of schema transformation that hides fields

**Reply by knotenpunkt** [reply]
ah ok thank you.

But what is then the difference between using the filed.transformer inside the SchemaDirectiveWiring VS SchemaTransformer (as used in the groovy-test-example you linked).

Why it isnt a good idea to hide/deactivate the fields directly in the SchemaDirectiveWiring

```
   SchemaGenerator schemaGenerator = new SchemaGenerator();
        GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeRegistry, buildWiring());


        SchemaTransformer schemaTransformer = new SchemaTransformer();

        GraphQLSchema newSchema = schemaTransformer.transform(graphQLSchema, new GraphQLTypeVisitorStub()
        {

            @Override
            public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context)
            {


                if (fieldDefinition.getDirective("blocked") !=null)
                {
                    //var changedNode = fieldDefinition.transform(builder -> {builder.name("barChanged");});
                    //return changeNode(context, changedNode);
                    return deleteNode(context);
                }
                return TraversalControl.CONTINUE;
            }
        });
```

**Reply by bbakerman** [reply]
> Why it isnt a good idea to hide/deactivate the fields directly in the SchemaDirectiveWiring

because SchemaDirectiveWiring does not support the idea of deleting (or adding) fields.  It is provided as a way to "tweak" a schema element as it is built.

SchemaTransformer takes a complete and valid schema and allows you to transform it.  They are similar but slightly different activities.

(Also SchemaDirectiveWiring predates SchemaTransformer so if we had our time again perhaps we would have combined them but they do not work like that)

**Reply by knotenpunkt** [reply]
ah ok, thank you, no more questions..... for the moment^^

**Reply by knotenpunkt** [reply]
ok and now i have the next question, its related to what we already discussed:

Where should i put the code, which not deletes, but creates new nodes? - Its not based on directives, but on the position in the sdl self.

So in that following example i want to duplicate (but with other datafetchers) all methods/fields which belongs to query and mutation

I also did this in the transformer-visitor:

```
 @Override
    public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context)
    {

        if (!(context.getParentContext().getLocation().getName().equals("query") || context.getParentContext().getLocation().getName().equals("mutation")))
        {
            return TraversalControl.CONTINUE;
        }

        insertAfter(context,fieldDefinition.transform(builder ->{
            builder.name(fieldDefinition.getName() + "AsClusterCall");
        }));


   //but where to put the wiring code?

  return TraversalControl.CONTINUE;

```

So the question is: is that ok? are there any better location?
Where i can wire the datafetchers or override them?

There is a builder method which is depreacted, there is the note to the Coderegistry. So im not sure if the transformer is the right location at all, creating new nodes (which other datafetchers?)

**Reply by knotenpunkt** [reply]
And one best practise question to the topic related blocking:

i want a blocking method for all, which a directive can have,
so i thought i take your method on which all other delegate to and check with an instanceof if the node is directive-holdable^^

But is that a good practice that way?


Same question to the DirectiveWiringClass: There you dont have such nice delegating-all method, how could i do something similar there, not deleting nodes but things for that the DirectiveWiringClass is for in a generic way?



```
 @Override
    protected TraversalControl visitGraphQLType(GraphQLSchemaElement node, TraverserContext<GraphQLSchemaElement> context)
    {

        if (node instanceof GraphQLDirectiveContainer)
        {
            if (((GraphQLDirectiveContainer) node).getDirective(this.blockedDirective) != null)
            {
                return deleteNode(context);
            }
        }
        return TraversalControl.CONTINUE;
    }
```

**Reply by bbakerman** [reply]
If you want to add new fields and new DataFetchers for them then you need to insert the types in the visitor and create a new CodeRegistry based on the old one

See graphql.schema.GraphQLCodeRegistry#newCodeRegistry(graphql.schema.GraphQLCodeRegistry)



Create one of those before you visit the schema, then build it up with new DFs and then build it and set it back into the resultant schema

```
GraphQLCodeRegistry.Builder codeRegistryBuilder = newCodeRegistry(startingSchema.getCodeRegistry)

GraphqlSchema newSchemaStep1 = transformSchema(startingSchema, new MyVisitor(codeRegistryBuilder));

GraphQLCodeRegistry newCodeRegistry = codeRegistryBuilder.build();

GraphqlSchema newSchemaStep2 = newSchemaStep1.transform( bld -> bld.codeRegistry(newCodeRegistry));

```

(That might not compile - I just pseudo wrote that)

**Reply by knotenpunkt** [reply]
ok that seems to be a solution?!

But why you splitted it, the codeRegistry from the fieldDefintiion? (its a unrelated question^^)

And how could i get for

`coordinates(GraphQLFieldsContainer parentType, GraphQLFieldDefinition fieldDefinition)`

the **parentType** inside 

` public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context)`

if i used there the `insertAfter(context, fieldDefinition.transform(builder ->`

it seems the fieldDefinition itself dont hold that information, and i dont have an idea how to get it from the context.
The result from `insertAfter` seems also not be helpful


And if i added here code, datafetchers and so on dynamically, does the DirectiveWiring catch up also the directives i added here dynamically?

**Reply by knotenpunkt** [reply]
ok as i figured it out i can get the parentType with 

`GraphQLFieldsContainer parentContainerType = (GraphQLFieldsContainer) context.getParentContext().thisNode();`

But i also figured out that first the directiveWiring is done and after that the visitor-code.
Is that possible to change there the ordering?
